openshift/console
# ODC-6712 & ODC-6714: Add a build decorator for Shipwright Builds and builds section into topology sidebar for Shipwright Builds and BuildRuns

**Fixes**: 

- [ODC-6712: Add a build decorator for Shipwright Builds (latest BuildRun) in topology graph](https://issues.redhat.com/browse/ODC-6712)
- [ODC-6714: Add builds section into topology sidebar (details tab) for Shipwright Builds and BuildRuns](https://issues.redhat.com/browse/ODC-6714)

**Solution Description**: 

1. Added a Decorator to the Shipwright-build-controller Deployment that would provide the status of the latest Build Run.
2. When a user clicks on the decorator, they would be navigated to the BuildRun logs page.
3. Added BuildRun section to the existing Topology Sidebar below the Pods Section, which lists all the Shipwright Builds.
4. Clicking on the Build Names redirects the user to the corresponding Builds Page.
5. If there are Buildruns corresponding to a build, they are listed below the corresponding Build along with the View Logs link that takes the user to the logs of the corresponding BuildRun.
6. The user is also given the facility to Start a Build from the Sidebar by clicking the Start Build button. If there are preexisting buildRuns to a Build, then the user is given the facility to re-run the latest BuildRun.


**Screen shots / Gifs for design review**: 

**[ODC-6712](https://issues.redhat.com//browse/ODC-6712): Add a build decorator for Shipwright Builds (latest BuildRun) in topology graph**

https://user-images.githubusercontent.com/47265560/181737247-43250389-c700-415c-895f-d19ad017234c.mp4

<br> 

**[ODC-6714](https://issues.redhat.com//browse/ODC-6714): Add builds section into topology sidebar (details tab) for Shipwright Builds and BuildRuns**

https://user-images.githubusercontent.com/47265560/181735055-76eb89d1-b5c2-473a-a56b-8b7bff4083ad.mp4


**E2E test coverage report**: 

![e2e-results](https://user-images.githubusercontent.com/47265560/184265095-381c3a03-9814-4a0a-972c-7750cfab681b.png)

![e2e-test-dec-side](https://user-images.githubusercontent.com/47265560/184265165-a2b514e1-cc31-4707-90cf-3391dd1b31b5.png)





**Test setup:**

1. Install the **Shipwright operator** along with the **OC pipelines** operator.
2. Create a **Shipwright Build** instance.
3. Navigate to the **samples** folder in the **shipwright-plugin package** and use the `oc create -f .` to install the resources in the namespace.
4. Come to the **Topology** page in the `build-examples` namespace.
5. Here, you will see various deployments/configs/k-servings created and will also the status of the BuildRun from the decorator attached to each of the resources. 
6. Click on each of the **deployments/configs/k-servings** to see the Topology Sidebar. You can see the new **BuildRuns** section below the Pods Section.
7. You can Start a new BuildRun or run the latest BuildRun from there.


**Browser conformance**:
